File: backend/src/main/java/vn/viettel/quanlycongno/service/MetricService.java
package vn.viettel.quanlycongno.service;

import vn.viettel.quanlycongno.dto.stats.StaffStatDTO;
import java.util.List;

/**
 * Service interface for retrieving metrics related to staff performance.
 * This service provides methods to get statistics about staff members based on their contracts and imports.
 */
public interface MetricService {
    /**
     * Get top staff with most contracts and imports for the past month
     * @param limit The number of staff to return
     * @return List of staff statistics
     */
    List<StaffStatDTO> getTopStaffForMonth(int limit);
    
    /**
     * Get top staff with most contracts and imports all time
     * @param limit The number of staff to return
     * @return List of staff statistics
     */
    List<StaffStatDTO> getTopStaffAllTime(int limit);

    /**
     * Get total count of contracts for a specific period
     * @param startDate The start date for filtering (inclusive)
     * @param endDate The end date for filtering (exclusive)
     * @return Total count of contracts
     */
    Long getTotalContractsForPeriod(String startDate, String endDate);

    /**
     * Get total count of invoices for a specific period
     * @param startDate The start date for filtering (inclusive)
     * @param endDate The end date for filtering (exclusive)
     * @return Total count of invoices
     */
    Long getTotalInvoicesForPeriod(String startDate, String endDate);
}

File: frontend/src/app/features/dashboard/services/metric.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse } from '../../../core/models/api-response.model';
import { StaffStatDTO } from '../../../shared/models/staff-stats.model';
import { environment } from '../../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class StaffStatsService {
  private http = inject(HttpClient);

  private readonly baseUrl = `${environment.apiUrl}/stats`;

  getTopStaffForMonth(limit = 10): Observable<ApiResponse<StaffStatDTO[]>> {
    const params = new HttpParams().set('limit', limit);
    return this.http.get<ApiResponse<StaffStatDTO[]>>(`${this.baseUrl}/staff/month`, { params });
  }

  getTopStaffAllTime(limit = 10): Observable<ApiResponse<StaffStatDTO[]>> {
    const params = new HttpParams().set('limit', limit);
    return this.http.get<ApiResponse<StaffStatDTO[]>>(`${this.baseUrl}/staff/all-time`, { params });
  }

  getTotalContractsForPeriod(startDate: string, endDate: string): Observable<ApiResponse<number>> {
    const params = new HttpParams()
      .set('startDate', startDate)
      .set('endDate', endDate);
    return this.http.get<ApiResponse<number>>(`${this.baseUrl}/contracts/period`, { params });
  }

  getTotalInvoicesForPeriod(startDate: string, endDate: string): Observable<ApiResponse<number>> {
    const params = new HttpParams()
      .set('startDate', startDate)
      .set('endDate', endDate);
    return this.http.get<ApiResponse<number>>(`${this.baseUrl}/invoices/period`, { params });
  }
}
